**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант № 7

Набор данных: [credit_risk.csv](credit_risk.csv)

Предметная область: Анализ кредитных рисков

Данный набор данных содержит информацию о кредитных заявках клиентов банка. Каждая строка представляет отдельную кредитную заявку с различными характеристиками заемщика и параметрами кредита.

Атрибуты:
- возраст
- доход
- статус домовладения (Собственный, Ипотека, Аренда)
- стаж работы в годах
- цель кредита
- сумма кредита по заявке
- процентная ставка
- статус одобрения кредита
- сумма кредита в процентах от дохода
- не допустил ли заявитель ранее дефолта по кредиту (да, нет)
- длина кредитной истории заявител


### 1.Чтение файла (набора данных)

In [893]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
import numpy as np
import warnings

# выключаем назойливые предупреждения об изменении API
warnings.filterwarnings("ignore", category=FutureWarning)

# чтение файла с помощью pandas
df = pd.read_csv('credit_risk.csv')

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [894]:
# применить метод head
df.head(20)

,Id,Age,Income,Home,Emp_length,Intent,Amount,Rate,Status,Percent_income,Default,Cred_length
0,0,22.0,59000,RENT,123.0,PERSONAL,35000,16.02,1,0.59,Y,3
1,1,21.0,9600,OWN,5.0,EDUCATION,1000,11.14,0,0.10,N,2
2,2,25.0,9600,MORTGAGE,1.0,MEDICAL,5500,12.87,1,0.57,N,3
3,3,23.0,65500,RENT,4.0,MEDICAL,35000,15.23,1,0.53,N,2
4,4,24.0,54400,RENT,8.0,MEDICAL,35000,14.27,1,0.55,Y,4
5,5,21.0,9900,OWN,2.0,VENTURE,2500,7.14,1,0.25,N,2
6,6,26.0,77100,RENET,8.0,EDUCATION,35000,12.42,1,0.45,N,3
7,7,24.0,78956,RENT,5.0,MEDICAL,35000,11.11,1,0.44,N,4
8,8,24.0,83000,RENT,8.0,PERSONAL,35000,8.90,1,0.42,N,2
9,9,21.0,10000,OWN,6.0,VENTURE,1600,14.74,1,0.16,N,3


2.2 Оценка данных с помощью метода info.

In [895]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              652 non-null    int64  
 1   Age             652 non-null    float64
 2   Income          652 non-null    int64  
 3   Home            652 non-null    object 
 4   Emp_length      643 non-null    float64
 5   Intent          652 non-null    object 
 6   Amount          652 non-null    int64  
 7   Rate            586 non-null    float64
 8   Status          652 non-null    int64  
 9   Percent_income  652 non-null    float64
 10  Default         652 non-null    object 
 11  Cred_length     652 non-null    int64  
dtypes: float64(4), int64(5), object(3)
memory usage: 61.3+ KB


2.3 Оценка данных с помощью метода describe.

In [896]:
# оцените числовые столбцы с помощью describe
df.describe()

,Id,Age,Income,Emp_length,Amount,Rate,Status,Percent_income,Cred_length
count,652.000000,652.000000,652.000000,643.000000,652.000000,586.000000,652.000000,652.000000,652.000000
mean,325.495399,24.285276,90008.855828,4.601866,18800.996933,12.289471,0.601227,0.279831,3.012270
std,188.352535,7.841061,69530.312525,7.281616,9152.896250,3.272772,0.490022,0.150231,0.808843
min,0.000000,21.000000,9600.000000,0.000000,1000.000000,5.420000,0.000000,0.010000,2.000000
25%,162.750000,23.000000,44000.000000,2.000000,10000.000000,10.250000,0.000000,0.160000,2.000000
50%,325.500000,24.000000,69998.000000,4.000000,21850.000000,12.180000,1.000000,0.280000,3.000000
75%,488.250000,25.000000,128499.000000,7.000000,25000.000000,14.720000,1.000000,0.380000,4.000000
max,649.000000,144.000000,500000.000000,123.000000,35000.000000,21.210000,1.000000,0.830000,4.000000


### Анализ результатов

#### Атрибуты

*   **Id (int64):** Уникальный идентификатор строки. Не несет аналитической ценности, так как дублирует стандартный индекс датафрейма.
*   **Age (float64):** Возраст заемщика. Некорректный тип данных – возраст должен быть целочисленным. Тип `float64` указывает на возможное наличие пропусков.
*   **Income (int64):** Годовой доход заемщика. Тип данных соответствует смыслу.
*   **Home (object):** Категориальный признак, описывающий статус владения жильем (`RENT`, `OWN`, `MORTGAGE`). Тип `object` (строка) является корректным.
*   **Emp_length (float64):** Стаж работы в годах. Как и возраст, должен быть целочисленным. Тип `float64` также свидетельствует о наличии пропусков.
*   **Intent (object):** Категориальный признак, описывающий цель кредита. Тип `object` корректен.
*   **Amount (int64):** Сумма кредита. Тип данных соответствует смыслу.
*   **Rate (float64):** Процентная ставка по кредиту. Тип данных `float64` (вещественное число) корректен.
*   **Status (int64):** Статус одобрения кредита (бинарный признак, где 0 — одобрено, 1 — не одобрено). Хотя тип int64 является технически корректным, для бинарного признака семантически более подходящим и эффективным с точки зрения использования памяти был бы тип bool (True/False). Этот тип данных также следует изменить в ходе предобработки.
*   **Percent_income (float64):** Доля дохода, идущая на погашение кредита. Тип `float64` корректен.
*   **Default (object):** Категориальный признак, указывающий на наличие дефолтов в прошлом (`Y` - да, `N` - нет). Тип `object` корректен.
*   **Cred_length (int64):** Длина кредитной истории в годах. Тип данных соответствует смыслу.

#### Выявленные проблемы и аномалии:

1.  **Пропущенные значения:** Метод `info()` показал наличие пропусков в двух ключевых столбцах:
    *   `Emp_length`: 9 пропущенных значений.
    *   `Rate`: 66 пропущенных значений.
    Эти пропуски необходимо обработать перед дальнейшим анализом.

2.  **Аномальные значения (выбросы):** Метод `describe()` и `head()` выявили нереалистичные максимальные значения:
    *   В столбце `Age` максимальное значение равно **144** года.
    *   В столбце `Emp_length` максимальное значение равно **123** года.
    Эти значения являются явными ошибками в данных и требуют исправления, так как они существенно искажают статистические метрики (особенно среднее значение).

3.  **Неявные дубликаты:** При просмотре первых строк (`head()`) в столбце `Home` обнаружено значение `RENET`, которое, вероятнее всего, является опечаткой и должно быть `RENT`. Это требует проверки и стандартизации.

4.  **Неоптимальные названия столбцов:** Названия, такие как `Emp_length` и `Cred_length`, используют смешанный регистр (`CamelCase`). Для удобства работы и соблюдения единого стиля лучше будет привести все названия к `snake_case` (например, `emp_length`). Кроме того, столбец `Id` является избыточным.

На основе данного анализа необходимо провести очистку и предобработку данных перед тем, как приступать к выполнению основных заданий.

 2.4 Оценка названий столбцов

In [897]:
# Вывести на экран названия столбцов с помощью df.columns
df.columns

Index(['Id', 'Age', 'Income', 'Home', 'Emp_length', 'Intent', 'Amount', 'Rate',
       'Status', 'Percent_income', 'Default', 'Cred_length'],
      dtype='object')

 2.5 Исправления столбцов

In [898]:
# Удаление столбца 'Id', так как он дублирует индекс
df = df.drop('Id', axis=1)

# Переименование остальных столбцов в snake case
df = df.rename(columns={
    'Age': 'age',
    'Income': 'income',
    'Home': 'home_ownership',
    'Emp_length': 'employment_length',
    'Intent': 'intent',
    'Amount': 'amount',
    'Rate': 'rate',
    'Status': 'status',
    'Percent_income': 'percent_income',
    'Default': 'has_previous_default',
    'Cred_length': 'credit_history_length'
})

# Вывод новых столбцов
df.columns

Index(['age', 'income', 'home_ownership', 'employment_length', 'intent',
       'amount', 'rate', 'status', 'percent_income', 'has_previous_default',
       'credit_history_length'],
      dtype='object')

### 3. Проверка пропусков

In [899]:
# Посчитаем, сколько строк останется после удаления всех строк с пропусками
rows_after_dropna = df.dropna().shape[0]

# Исходное количество строк
initial_rows = df.shape[0]

# Количество удаленных строк
deleted_rows = initial_rows - rows_after_dropna

# Процент потерянных данных
loss_percentage = (deleted_rows / initial_rows) * 100

print(f"Исходное количество строк: {initial_rows}")
print(f"Количество строк после удаления пропусков: {rows_after_dropna}")
print(f"Будет удалено строк: {deleted_rows}")
print(f"Процент потерь: {loss_percentage:.2f}%")

Исходное количество строк: 652
Количество строк после удаления пропусков: 578
Будет удалено строк: 74
Процент потерь: 11.35%


In [900]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).
print('Общее количество пропусков')
print(df.isna().sum())

# Замена аномальных значений на NaN
df.loc[df['age'] > 100, 'age'] = np.nan
df.loc[df['employment_length'] > 60, 'employment_length'] = np.nan

print("\nКоличество пропусков после замены аномалий:")
print(df.isna().sum())

print('\n Описательная статистика после замены аномалий:')
print(df.describe())

# Расчет медианных значений
median_age = df['age'].median()
median_employment_length = df['employment_length'].median()
median_rate = df['rate'].median()

# Заменяем все NaN на медианные значения
df = df.fillna({'age': median_age})
df = df.fillna({'employment_length': median_employment_length})
df = df.fillna({'rate': median_rate})

print("\nКоличество пропусков после замены NaN на медианные значения:")
print(df.isna().sum())
print('\n Описательная статистика после замены NaN на медианные значения:')
print(df.describe())


Общее количество пропусков
age                       0
income                    0
home_ownership            0
employment_length         9
intent                    0
amount                    0
rate                     66
status                    0
percent_income            0
has_previous_default      0
credit_history_length     0
dtype: int64

Количество пропусков после замены аномалий:
age                       3
income                    0
home_ownership            0
employment_length        11
intent                    0
amount                    0
rate                     66
status                    0
percent_income            0
has_previous_default      0
credit_history_length     0
dtype: int64

 Описательная статистика после замены аномалий:
              age         income  employment_length        amount        rate  \
count  649.000000     652.000000         641.000000    652.000000  586.000000   
mean    23.764253   90008.855828           4.232449  18800.996933   12.2894

### Обоснование выбранной стратегии обработки пропусков

В ходе анализа данных были выявлены пропуски в столбцах `employment_length` и `loan_rate`, а также аномальные значения в `person_age` и `employment_length`, которые были преобразованы в пропуски.

**Выбранная стратегия:** Заполнение пропусков **медианным** значением.

**Аргументация:**

1.  **Сохранение данных:** Удаление строк с пропусками привело бы к потере значительной части данных (более 11%), что могло бы исказить результаты. Поэтому был выбран метод заполнения.
2.  **Надежность:** Для числовых признаков был выбран метод заполнения медианой, так как она, в отличие от среднего арифметического, устойчива к выбросам и экстремальным значениям. Это гарантирует, что пропуски заменяются наиболее "типичным" значением для большинства данных, а не смещенным из-за аномалий.

### Альтернативные способы работы с пропусками

1.  **Удаление строк или столбцов (`.dropna()`):** Простой метод, который подходит, если пропусков очень мало. В нашем случае это привело бы к значительной потере данных.
2.  **Заполнение средним значением (`.fillna(df['column'].mean())`):** Подходит для числовых данных без сильных выбросов. В нашем наборе данных этот метод был бы менее точным из-за наличия аномально высоких значений.
3.  **Заполнение нулем или другим значением (`.fillna(0)`):** Применяется, когда можно логически предположить, что пропуск означает отсутствие чего-либо (например, стажа).

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [901]:
# Считаем количество полностью одинаковых строк в датафрейме.
print(f"Количество явных дубликатов до удаления: {df.duplicated().sum()}")

# Удаляем дубликаты
df.drop_duplicates(inplace=True)

# Обновляем индексы, удаляя из них 'дыры', которые могли появиться после удаления дубликатов.
# Также запрещаем создавать столбец 'Index' из удаленных индексов.
df.reset_index(drop=True, inplace=True)

print(f"Количество явных дубликатов после удаления: {df.duplicated().sum()}")
print(f"Новый размер датафрейма: {df.shape}")

Количество явных дубликатов до удаления: 2
Количество явных дубликатов после удаления: 0
Новый размер датафрейма: (650, 11)


#### Проверка неявных дубликатов

In [902]:
print("Уникальные значения в 'home_ownership':")
print(df['home_ownership'].unique())

print("\nУникальные значения в 'intent':")
print(df['intent'].unique())

print("\nУникальные значения в 'has_previous_default':")
print(df['has_previous_default'].unique())

Уникальные значения в 'home_ownership':
['RENT' 'OWN' 'MORTGAGE' 'RENET' 'OTHER']

Уникальные значения в 'intent':
['PERSONAL' 'EDUCATION' 'MEDICAL' 'VENTURE' 'HOMEIMPROVEMENT'
 'DEBTCONSOLIDATION']

Уникальные значения в 'has_previous_default':
['Y' 'N' 'No']


In [903]:
# Чтобы принять обоснованное решение об удалении категории 'OTHER',
# посчитаем, сколько раз встречается каждое значение в столбце 'home_ownership'.
# Метод .value_counts() идеально подходит для этой задачи.

print("Количество записей для каждой категории в 'home_ownership':")
print(df['home_ownership'].value_counts())

Количество записей для каждой категории в 'home_ownership':
home_ownership
RENT        413
MORTGAGE    136
OWN          98
OTHER         2
RENET         1
Name: count, dtype: int64


In [904]:
# Заменяем 'RENET' на 'RENT'
df.replace({'home_ownership': 'RENET'}, 'RENT', inplace=True)

# Заменяем 'No' на 'N'
df.replace({'has_previous_default': 'No'}, 'N', inplace=True) 

# Удаляем строки с категорией 'OTHER', так как их всего две.
df = df[df['home_ownership'] != 'OTHER']

# Сбрасываем индекс, так как удалили строки
df.reset_index(drop=True, inplace=True)


print("Уникальные значения в 'home_ownership' после всех исправлений:")
print(df['home_ownership'].unique())

print("\nУникальные значения в 'has_previous_default' после всех исправлений:")
print(df['has_previous_default'].unique())

print(f"\nИтоговый размер датафрейма после очистки неявных дубликатов: {df.shape}")


Уникальные значения в 'home_ownership' после всех исправлений:
['RENT' 'OWN' 'MORTGAGE']

Уникальные значения в 'has_previous_default' после всех исправлений:
['Y' 'N']

Итоговый размер датафрейма после очистки неявных дубликатов: (648, 11)


### Обоснование действий по обработке дубликатов

Были проверены и устранены явные и неявные дубликаты для обеспечения корректности анализа.

1.  **Явные дубликаты:**
    *   **Проблема:** Обнаружены 2 полностью идентичные строки.
    *   **Действие:** Удалены с помощью `df.drop_duplicates()`.
    *   **Причина:** Такие строки не несут новой информации и искажают статистику.

2.  **Неявные дубликаты:**
    *   **Проблема 1:** В столбце `home_ownership` найдена опечатка `'RENET'`.
    *   **Действие 1:** Исправлена на `'RENT'` с помощью `.replace()`.

    *   **Проблема 2:** В столбце `has_previous_default` найден синоним `'No'` для значения `'N'`.
    *   **Действие 2:** Значение `'No'` заменено на `'N'` для стандартизации.

    *   **Проблема 3:** Категория `'OTHER'` в `home_ownership` содержит всего 2 записи.
    *   **Действие 3:** Строки с этой категорией были удалены, так как представляют собой статистически незначимый "шум".

    *   **Причина (общая):** Устранение неявных дубликатов обеспечивает правильную группировку данных, так как семантически одинаковые значения теперь обрабатываются как одна категория.

### 5. Провека типов данных

In [905]:
# Проверяем уникальные значения в 'status'
print("Уникальные значения в 'status':")
print(df['status'].unique())

Уникальные значения в 'status':
[1 0]


In [906]:
# Преобразуем числовые столбцы из float64 в int64
df['age'] = df['age'].astype('int64')
df['employment_length'] = df['employment_length'].astype('int64')

# Преобразуем бинарные столбцы в bool
df['status'] = df['status'].astype('bool')
df['has_previous_default'] = (df['has_previous_default'] == 'Y')

print("Типы данных после всех преобразований:")
df.info()

Типы данных после всех преобразований:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 648 entries, 0 to 647
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   age                    648 non-null    int64  
 1   income                 648 non-null    int64  
 2   home_ownership         648 non-null    object 
 3   employment_length      648 non-null    int64  
 4   intent                 648 non-null    object 
 5   amount                 648 non-null    int64  
 6   rate                   648 non-null    float64
 7   status                 648 non-null    bool   
 8   percent_income         648 non-null    float64
 9   has_previous_default   648 non-null    bool   
 10  credit_history_length  648 non-null    int64  
dtypes: bool(2), float64(2), int64(5), object(2)
memory usage: 47.0+ KB


### Обоснование преобразования типов данных

Было выполнено приведение типов столбцов к их правильному семантическому виду.

1.  **Преобразование `float64` в `int64`:**
    *   **Столбцы:** `age`, `employment_length`.
    *   **Причина:** Возраст и стаж являются целочисленными.

2.  **Преобразование бинарных признаков в `bool`:**
    *   **Столбцы:** `status` (из `int64`) и `has_previous_default` (из `object`).
    *   **Причина:** Тип `bool` (`True`/`False`) семантически точнее отражает суть бинарных данных ("да/нет") и более эффективен по использованию памяти. Перед преобразованием `status` была проведена проверка, которая подтвердила отсутствие в столбце значений, отличных от `0` и `1`.

### 6. Группировка данных

#### Задание 1

Группировка - цель кредита (intent) по количеству статуса домовладения (home).

In [907]:
# Группируем данные по цели кредита ('intent'),
# а затем для каждой цели считаем количество вхождений каждого статуса домовладения ('home_ownership').
task1_result = df.groupby('intent')['home_ownership'].value_counts()

# Выводим результат
task1_result

intent             home_ownership
DEBTCONSOLIDATION  RENT              72
                   MORTGAGE          21
                   OWN                9
EDUCATION          RENT              94
                   MORTGAGE          34
                   OWN               16
HOMEIMPROVEMENT    RENT              39
                   OWN               16
                   MORTGAGE          13
MEDICAL            RENT              70
                   MORTGAGE          22
                   OWN               15
PERSONAL           RENT              69
                   OWN               22
                   MORTGAGE          19
VENTURE            RENT              70
                   MORTGAGE          27
                   OWN               20
Name: count, dtype: int64

Результат группировки показывает распределение статуса владения жильем для каждой цели кредита. Можно сделать следующие выводы:

1.  **Аренда (`RENT`) доминирует:** Для всех без исключения целей кредитования наибольшее число заявок поступает от клиентов, арендующих жилье. Это самая крупная и активная группа заемщиков. Например, для цели `EDUCATION` от арендаторов поступило 94 заявки, что значительно превышает другие категории.
2.  **Нелогичное поведение для "Улучшения жилья":** Неожиданно, что даже для кредитов на улучшение жилья (`HOMEIMPROVEMENT`) больше всего заявок от арендаторов (39), а не от владельцев (`OWN` - 16) или ипотечников (`MORTGAGE` - 13). Это может указывать либо на ошибки в данных, либо на то, что заемщики неверно указывают цель кредита.
3.  **Собственники (`OWN`) в меньшинстве:** Клиенты, полностью владеющие жильем, стабильно составляют наименьшую или вторую по численности группу заемщиков по всем целям кредитования. Это может говорить об их большей финансовой независимости.

#### Задание 2

Группировка - статус домовладения (home) по количеству дефолтов (default). Создать датафрейм. Переименовать столбец с количеством в "count". Отсортировать по убыванию столбца "count".

In [908]:
# 1. Группируем по 'home_ownership' и для каждой группы считаем количество True/False в 'has_previous_default'.
# 2. .reset_index(name='count') преобразует результат (Series) в DataFrame
#    и дает столбцу со значениями имя 'count'.
# 3. .sort_values(by='count', ascending=False) сортирует полученный DataFrame по убыванию количества.
task2_result = df.groupby('home_ownership')['has_previous_default'].value_counts()
task2_result = task2_result.reset_index(name='count')
task2_result.sort_values(by='count', ascending=False, inplace=True)

# Выводим итоговый датафрейм
task2_result


,home_ownership,has_previous_default,count
4,RENT,False,308
0,MORTGAGE,False,107
5,RENT,True,106
2,OWN,False,85
1,MORTGAGE,True,29
3,OWN,True,13


Данная группировка показывает абсолютное распределение исторических дефолтов среди клиентов с разным статусом владения жильем.

1.  **Лидерство арендаторов:** Арендаторы (`RENT`) составляют самую большую группу заемщиков как среди тех, у кого не было дефолтов (308 заявок), так и среди тех, у кого они были (106 заявок).
2.  **Дефолты - не редкость:** В целом, у значительной части клиентов (`106 + 29 + 13 = 148` из 648) в прошлом были дефолты.
3.  **Доля дефолтов у арендаторов:** У категории `RENT` наблюдается самое высокое абсолютное число клиентов с предыдущими дефолтами (106). Это может указывать на более высокий кредитный риск в данной группе по сравнению с владельцами жилья или плательщиками ипотеки.

#### Задание 3

Сводная таблица (pivot_table) - средняя сумма кредита (amount) для каждой цели кредита. Отсортировать по убыванию. Округлить до трёх знаков.

In [909]:
# Создаем сводную таблицу.
# index='intent': строки таблицы будут соответствовать целям кредита.
# values='amount': значения в ячейках - это суммы кредита.
# aggfunc='mean': функция агрегации - вычисление среднего значения.
task3_pivot = df.pivot_table(index='intent', values='amount', aggfunc='mean')

# Сортируем результат по убыванию средней суммы кредита
task3_pivot.sort_values(by='amount', ascending=False, inplace=True)

# Округляем значения до трех знаков после запятой
task3_pivot = task3_pivot.round(3)

# Выводим итоговую таблицу
task3_pivot


,amount
intent,
DEBTCONSOLIDATION,19578.186
PERSONAL,19458.409
EDUCATION,19264.757
MEDICAL,18732.243
HOMEIMPROVEMENT,17884.559
VENTURE,17732.692


Сводная таблица показывает среднюю запрашиваемую сумму кредита в зависимости от его цели.

1.  **Высокие суммы по всем категориям:** Средняя запрашиваемая сумма кредита является довольно высокой (от ~17 700 до ~19 500) и относительно близкой для всех целей кредитования. Это может указывать на то, что банк, чьи данные анализируются, специализируется на выдаче крупных займов (если сумма конечно в долларах).
2.  **Лидеры по сумме:** Наибольшая средняя сумма запрашивается на **консолидацию долга** (`DEBTCONSOLIDATION`) и на **личные цели** (`PERSONAL`). Высокая сумма для личных целей может говорить о том, что под этой категорией скрываются крупные покупки, а не мелкие потребительские расходы.
3.  **Бизнес и улучшение жилья:** Кредиты на **бизнес-проекты** (`VENTURE`) и **улучшение жилья** (`HOMEIMPROVEMENT`) имеют самые низкие средние суммы, хотя и ненамного. Это может показаться нетипичным, так как эти цели часто требуют значительных вложений.

#### Задание 4

Сводная таблица (pivot_table) - средний доход (income) для каждого статуса домовладения (home) - строки и цели кредита - столбцы. Отсортировать по возрастанию home. Округлить до двух знаков.

In [910]:
# Создаем сводную таблицу
# index='home_ownership': строки таблицы
# columns='intent': столбцы таблицы
# values='income': значения для агрегации (доход)
# aggfunc='mean': функция агрегации (среднее)
task4_pivot = df.pivot_table(index='home_ownership',
                             columns='intent',
                             values='income',
                             aggfunc='mean')

# Сортируем по индексу (home_ownership) в алфавитном порядке и округляем значения
task4_pivot = task4_pivot.sort_index(ascending=True).round(2)

# Выводим итоговую таблицу
task4_pivot


intent,DEBTCONSOLIDATION,EDUCATION,HOMEIMPROVEMENT,MEDICAL,PERSONAL,VENTURE
home_ownership,,,,,,
MORTGAGE,128852.05,141393.41,144012.62,117061.55,102129.53,145813.37
OWN,14334.67,63824.75,72565.44,39385.07,69722.55,44181.00
RENT,85176.99,87780.28,83998.54,79992.63,86012.55,88363.53


Сводная таблица показывает средний доход заемщиков в разрезе статуса владения жильем и цели кредита. Анализ таблицы выявляет несколько ключевых и неожиданных закономерностей.

1.  **Клиенты с ипотекой (`MORTGAGE`) имеют самый высокий доход:** Это ожидаемо. В данной группе самый высокий средний доход (более 100 000 по всем категориям), что подтверждает их высокую платежеспособность. Наибольший средний доход у них наблюдается при взятии кредита на бизнес (`VENTURE`) и улучшение жилья (`HOMEIMPROVEMENT`).

2.  **Аномально низкий доход у владельцев (`OWN`):** Неожиданно, что клиенты, полностью владеющие своим жильем, показывают значительно более низкий средний доход, чем даже арендаторы. Особенно выделяется аномально низкое значение дохода в **14 334** для цели "консолидация долга". Это может указывать на несколько вещей:
    *   Возможно, в эту категорию попадают пенсионеры, владеющие жильем, но имеющие низкий текущий доход.
    *   Могут присутствовать ошибки в данных, которые не были устранены на предыдущих этапах.
    *   Либо это специфический сегмент клиентов, который требует отдельного, более глубокого изучения.

3.  **Арендаторы (`RENT`) занимают среднее положение:** Вопреки ожиданиям, арендаторы имеют стабильно более высокий средний доход (около 80-88 000), чем владельцы жилья, что делает их второй по уровню дохода группой после ипотечников.

### Вывод

В ходе лабораторной работы был проведен анализ и предобработка набора данных о кредитных заявках с целью подготовки их для дальнейшего анализа.

На этапе предобработки были устранены различные дефекты данных. В частности, были исправлены аномальные значения возраста (до **144** лет) и стажа (до **123** лет), а также заполнены **80** пропущенных значений в числовых столбцах с использованием медианы. Кроме того, были удалены явные дубликаты и исправлены неявные, возникшие из-за опечаток, что позволило обеспечить корректность категориальных данных.

Анализ очищенных данных показал, что самой активной группой заемщиков являются арендаторы (`RENT`), которые также лидируют по абсолютному числу прошлых дефолтов (**106** случаев). Средние суммы кредита по всем целям оказались высокими и близкими, находясь в диапазоне от **17 732** до **19 578** у.е.

Ключевой находкой стала четкая связь между доходом и статусом жилья. Заемщики с ипотекой (`MORTGAGE`) имеют самый высокий средний доход (более **102 000** у.е.). Неожиданно, владельцы собственного жилья (`OWN`) показали аномально низкий средний доход (например, всего **14 334** у.е. для цели "консолидация долга"), что даже ниже, чем у арендаторов (в среднем **86 000** у.е.). Эта особенность требует дополнительного изучения. Таким образом, предобработка данных позволила не только повысить их качество, но и выявить важные закономерности.

### Дополнительное задание

## Дополинтельное задание №2

Создать столбец "Категория дохода (Income)" с помощью категоризации, выделив 3 категории: низкий, средний и высокий доход. Фильтрацию границ категорий выбрать на основе описательной статистики данных. Затем построить сводную таблицу, показывающую средний и медианный доход для каждой категории в разрезе статуса домовладения (home_ownership).

In [911]:
# Выводим описательную статистику для столбца income
income_stats = df['income'].describe()
print("Описательная статистика дохода:")
income_stats

Описательная статистика дохода:


count       648.000000
mean      90335.779321
std       69598.362888
min        9600.000000
25%       44723.000000
50%       70000.000000
75%      129997.000000
max      500000.000000
Name: income, dtype: float64

In [912]:
# Определяем границы категорий на основне описательной статистике
# Хотя я бы лучше разбил на 0.33 и 0.66 перцентили, но раз задача стоит выполнить фильтрацию по 'описательной статистике'
# То сделаем по предоставленным данным из describe
# Можно было бы использовать еще и qcut
low_income = income_stats['25%']
high_income = income_stats['75%']
max_income = income_stats['max']

# Создаем категории
bins = [0, low_income, high_income, max_income]
print("\nГраницы категорий:")
print(bins)

labels = ['низкий', 'средний', 'высокий']

df['income_category'] = pd.cut(
    df['income'],
    bins=bins,
    labels=labels,
    include_lowest=True
)

print("\nРаспределение по категориям:")
print(df['income_category'].value_counts())

df[['income', 'income_category']]




Границы категорий:
[0, np.float64(44723.0), np.float64(129997.0), np.float64(500000.0)]

Распределение по категориям:
income_category
средний    324
низкий     162
высокий    162
Name: count, dtype: int64


,income,income_category
0,59000,средний
1,9600,низкий
2,9600,низкий
3,65500,средний
4,54400,средний
...,...,...
643,51012,средний
644,20004,низкий
645,20004,низкий
646,58000,средний


In [913]:
# Сводная таблица
income_pivot = df.pivot_table(
    index='income_category',
    columns='home_ownership',
    values='income',
    aggfunc=['mean', 'median'],
).round(2)

print("\nГраницы категорий:")
print(f"- Низкий доход: < {low_income:.0f}")
print(f"- Средний доход: {low_income:.0f}–{high_income:.0f}")
print(f"- Высокий доход: > {high_income:.0f}\n")

income_pivot



Границы категорий:
- Низкий доход: < 44723
- Средний доход: 44723–129997
- Высокий доход: > 129997



mean                          median                    
home_ownership    MORTGAGE        OWN       RENT  MORTGAGE       OWN      RENT
income_category                                                               
низкий            16951.61   15506.97   38637.53   18000.0   15360.0   39500.0
средний          125800.00        NaN   73436.29  125800.0       NaN   70000.0
высокий          200569.96  215499.79  176669.59  190650.0  189996.0  168600.0

In [914]:
# Проверка наличия комбинаций
print("Количество наблюдений по комбинациям:")
print(df.groupby(['income_category', 'home_ownership']).size())

Количество наблюдений по комбинациям:
income_category  home_ownership
низкий           MORTGAGE           51
                 OWN                79
                 RENT               32
средний          MORTGAGE            1
                 OWN                 0
                 RENT              323
высокий          MORTGAGE           84
                 OWN                19
                 RENT               59
dtype: int64


Созданная сводная таблица анализирует средний и медианный доход заемщиков, разделенных на три категории (низкий, средний, высокий), в разрезе статуса владения жильем. Анализ выявляет несколько примечательных закономерностей.

1.  **Ипотечные заемщики (`MORTGAGE`) - самые состоятельные.**
    Эта группа стабильно демонстрирует высокий доход во всех категориях. Даже в "низкой" категории их средний доход (**16 951 у.е.**) выше, чем у владельцев (`OWN`).

2.  **Арендаторы (`RENT`) — самый многочисленный сегмент среднего класса.**
    Ключевой особенностью является то, что почти все клиенты (**323 из 324**) со средним доходом (44 723–129 997 у.е.) являются именно арендаторами. Их средний доход в этой группе составляет **73 436 у.е.** Это может указывать на сегмент молодых специалистов, которые имеют стабильный доход, но еще не готовы к покупке собственного жилья.

3.  **Странности владельцев жилья (`OWN`)**
    Эта группа демонстрирует полярное распределение доходов: они либо имеют самый низкий средний доход (**15 506 у.е.**), либо один из самых высоких (**215 499 у.е.**). При этом в категории "средний доход" владельцы отсутствуют вовсе (**0 наблюдений**), что отражается как `NaN` в сводной таблице. Такая аномалия может быть вызвана наличием в этой группе например пенсионеров с низким текущим доходом, но владеющих недвижимостью.

## Дополинтельное задание №23

Отфильтровать набор данных. Выбрать топ 2 цели кредита по средней сумме
кредита + топ 2 статуса землевладения по медианной сумме кредита. Создать
сводные таблицы: цель кредита, статус домовладения - медианный, средний,
максимальный доход.

In [915]:
# Топ-2 цели кредита по средней сумме
mean_amount_by_intent = df.groupby('intent')['amount'].mean().sort_values(ascending=False)
print("Средняя сумма кредита по целям:")
print(mean_amount_by_intent)

top_intents = mean_amount_by_intent.index[:2].tolist()
print(f"\nОтобранные топ-2 цели кредита: {top_intents}\n")


# Топ-2 статуса владения по медианной сумме
median_amount_by_ownership = df.groupby('home_ownership')['amount'].median().sort_values(ascending=False)
print("Медианная сумма кредита по статусу владения:")
print(median_amount_by_ownership)

top_ownerships = median_amount_by_ownership.index[:2].tolist()
print(f"\nОтобранные топ-2 статуса владения: {top_ownerships}")


Средняя сумма кредита по целям:
intent
DEBTCONSOLIDATION    19578.186275
PERSONAL             19458.409091
EDUCATION            19264.756944
MEDICAL              18732.242991
HOMEIMPROVEMENT      17884.558824
VENTURE              17732.692308
Name: amount, dtype: float64

Отобранные топ-2 цели кредита: ['DEBTCONSOLIDATION', 'PERSONAL']

Медианная сумма кредита по статусу владения:
home_ownership
RENT        24000.0
MORTGAGE    10000.0
OWN          4100.0
Name: amount, dtype: float64

Отобранные топ-2 статуса владения: ['RENT', 'MORTGAGE']


In [916]:
# Фильтруем датафрейм по найденным критериям из предыдущей ячейки
filtered_df = df[
    (df['intent'].isin(top_intents)) &
    (df['home_ownership'].isin(top_ownerships))
]

print('Отфильтрованный датафрейм (head):')
print(filtered_df.head())

# Итоговая сводная таблица
income_analysis_pivot = filtered_df.pivot_table(
    index='intent',
    columns='home_ownership',
    values='income',
    aggfunc=['median', 'mean', 'max'],
).round(2)
print('Итоговая сводная таблица:')
income_analysis_pivot


Отфильтрованный датафрейм (head):
    age  income home_ownership  employment_length             intent  amount  \
0    22   59000           RENT                  4           PERSONAL   35000   
8    24   83000           RENT                  8           PERSONAL   35000   
15   23  500000       MORTGAGE                  7  DEBTCONSOLIDATION   30000   
18   23  113000           RENT                  8  DEBTCONSOLIDATION   35000   
21   25  137000           RENT                  9           PERSONAL   34800   

     rate  status  percent_income  has_previous_default  \
0   16.02    True            0.59                  True   
8    8.90    True            0.42                 False   
15  10.65   False            0.06                 False   
18  18.25    True            0.31                 False   
21  16.77   False            0.25                  True   

    credit_history_length income_category  
0                       3         средний  
8                       2         средний 

median                mean                max        
home_ownership     MORTGAGE     RENT   MORTGAGE      RENT MORTGAGE    RENT
intent                                                                    
DEBTCONSOLIDATION  160000.0  74034.0  128852.05  85176.99   500000  306000
PERSONAL           160000.0  77650.0  102129.53  86012.55   200000  280000

Анализ доходов топ-сегментов заемщиков (цели `DEBTCONSOLIDATION`, `PERSONAL` и статусы `MORTGAGE`, `RENT`) показал:

1.  **Заемщики с ипотекой (`MORTGAGE`) значительно состоятельнее.** Их медианный доход (**160 000 у.е.**) более чем в два раза превышает доход арендаторов, а максимальный доход достигает **500 000 у.е.**, что подтверждает их высокую платежеспособность.

2.  **Консолидация долга (`DEBTCONSOLIDATION`) — удел более обеспеченных.** Эта цель привлекает клиентов с более высоким средним доходом (**128 852 у.е.** у ипотечников) по сравнению с кредитами на личные цели. У арендаторов (`RENT`) такой ярко выраженной разницы в доходе между целями не наблюдается.